src/deepsparse/transformers/engines/nl_decoder_engine.py

```diff
@@ -0,0 +1,323 @@
+        self.sampling_temperature = sampling_temperature
+        self.deterministic = deterministic
+        self.input_ids_length = input_ids_length
+        self.kv_cache_enabled = kv_cache_enabled
+        self.kv_cache = (
+            DecoderKVCache(use_deepsparse_cache) if kv_cache_enabled else None
+        )
+        self._num_tokens = 0  # the number of tokens processed so far
+        self._freeze_first_position = self._should_freeze_first_position(tokenizer)
+        self._session_id = generate_session_id()
+
+    @property
+    def session_id(self) -> str:
+        """
+        :return: The session id for the kv_cache if enabled
+        """
+        return self._session_id
+
+    @session_id.setter
+    def session_id(self, session_id: str):
+        """
+        :param session_id: The session id to set for the kv_cache
+        """
+        self._session_id = session_id
+
+    @property
+    def onnx_input_names_no_cache(self) -> List[str]:
+        """
+        :return: The input names for the onnx model, excluding
+            the potential kv cache inputs
+        """
+        return [
+            name
+            for name in self.engine.input_names
+            if not name.startswith(_CACHE_INPUT_NAME)
+        ]
+
+    def __call__(
+        self,
+        inp: List[numpy.ndarray],
+        val_inp: bool = True,
+    ) -> Tuple[numpy.ndarray, numpy.ndarray]:
+        """
+        The main entry point for running the engine.
+
+        :param inp: The input to run the engine with. We expect a
+            list of numpy arrays that contain the input ids,
+            attention mask, and position ids (optionally)
+        :param val_inp: Whether the input is for validation or not
+        :return: The generated token and corresponding logits
+        """
+        if self.kv_cache:
+            # if kv cache is enabled, we need to add the kv cache state
+            # to the input
+            inp = self._add_kv_cache_to_input(inp)
+
+        out = self.engine.run(inp, val_inp)
+
+        self._num_tokens += self.input_ids_length
+
+        if self.kv_cache:
+            logits, *kv_cache_state = out
+            self._update_kv_cache(
+                kv_cache_state=kv_cache_state,
+            )
+        else:
+            logits = out[0]
+            B, S, V = logits.shape  # batch, sequence, vocab
+            logits = logits[:, -1, :].reshape(B, 1, V)  # only take the last token
+
+        token = self.generate_token(logits=logits)
+
+        return token, logits
+
+    def transfer_cache_state(self, engine: "NLDecoderEngine"):
+        """
+        Transfers the kv cache state and the number of tokens processed
+        information from another NLDecoderEngine. Call this method when
+        you want to transfer the kv cache state from one engine to another.
+
+        :param engine: The `NLDecoderEngine` to transfer the kv cache state
+            from
+        """
+        state = engine.kv_cache.cached_inputs
```

Comment:
let's have this function take the engine KVCache directly (again this is because we will have multiple cache objects per engine once we're in multi session/multi stream)

Reply:
Ok, but this only works nicely if KVCache tracks the amount of cache that it has processed.

Reply:
Agreement: each KVCache session tracks the number of processed tokens